File: code/php/sdk/src/ServerSentEventGenerator.php
<?php
/**
 * @copyright Copyright (c) PutYourLightsOn
 */

namespace starfederation\datastar;

use starfederation\datastar\events\EventInterface;
use starfederation\datastar\enums\ConsoleMode;
use starfederation\datastar\enums\EventType;
use starfederation\datastar\enums\FragmentMergeMode;
use starfederation\datastar\events\Console;
use starfederation\datastar\events\Remove;
use starfederation\datastar\events\Fragment;
use starfederation\datastar\events\Redirect;
use starfederation\datastar\events\Signal;

class ServerSentEventGenerator
{
    public function __construct()
    {
        $this->sendHeaders();
    }

    /**
     * Renders a fragment in the DOM.
     *
     * @param array{
     *     selector: string|null,
     *     mergeMode: FragmentMergeMode|null,
     *     settleDuration: int|null,
     *     useViewTransition: bool|null,
     *     eventId: string|null,
     *     retryDuration: int|null,
     * } $options
     */
    public function renderFragment(string $data, array $options = []): void
    {
        $this->sendEvent(new Fragment($data, $options));
    }

    /**
     * Removes one or more fragments from the DOM.
     *
     * @param array{
     *      eventId: string|null,
     *      retryDuration: int|null,
     *  } $options
     */
    public function removeFragments(string $selector, array $options = []): void
    {
        $this->sendEvent(new Remove($selector, $options));
    }

    /**
     * Updates values in the store.
     */
    public function patchStore(string $data, array $options = []): void
    {
        $this->sendEvent(new Signal($data, $options));
    }

    /**
     * Removes one or more paths from the store.
     */
    public function removeFromStore(array $paths): void
    {
        $this->sendEvent(new Remove(paths: $paths));
    }

    /**
     * Redirects the browser.
     */
    public function redirect(string $url): void
    {
        $this->sendEvent(new Redirect($url));
    }

    /**
     * Sends a message to the browser console.
     */
    public function console(ConsoleMode $mode, string $message): void
    {
        $this->sendEvent(new Console($mode, $message));
    }

    /**
     * Sends the response headers, if not already sent.
     */
    protected function sendHeaders(): void
    {
        if (headers_sent()) {
            return;
        }

        header('Content-Type: text/event-stream');
        header('Cache-Control: no-cache');
        header('Connection: keep-alive');

        // Disable buffering for Nginx
        // https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering
        header('X-Accel-Buffering: no');
    }

    /**
     * Sends an event.
     */
    protected function sendEvent(EventInterface $event): void
    {
        $this->send(
            $event->getEventType(),
            $event->getDataLines(),
        );
    }

    /**
     * Sends a Datastar event.
     *
     * @param EventType $eventType
     * @param string[] $dataLines
     * @param array{
     *     id: string|null,
     *     retry: int|null,
     * } $options
     */
    protected function send(EventType $eventType, array $dataLines, array $options = []): void
    {
        $eventData = new ServerSentEventData(
            $eventType,
            $dataLines,
            null,
            Defaults::DEFAULT_SSE_SEND_RETRY,
        );

        foreach ($options as $key => $value) {
            if (property_exists($eventData, $key)) {
                $eventData->$key = $value;
            }
        }

        $output = [
            'event: ' . $eventData->eventType->value,
            'id: ' . $eventData->eventId,
            'retry: ' . $eventData->retryDuration,
        ];

        foreach ($eventData->data as $line) {
            $output[] = 'data: ' . $line;
        }

        echo implode("\n", $output) . "\n\n";

        if (ob_get_contents()) {
          ob_end_flush();
        }
        flush();
    }
}

File: code/dotnet/sdk/src/ServerSentEventGenerator.fs
namespace StarFederation.Datastar

open System.Text
open System.Text.Json
open System.Threading.Tasks
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.Primitives
open Microsoft.Net.Http.Headers

type IServerSentEventContext =
    interface
        inherit ISendEvent
    end

type ServerSentEventGenerator (context:IServerSentEventContext) =
    new (httpContext:HttpContext) =
        ServerSentEventGenerator (ServerSentEventHttpContext httpContext)
    new (httpContextAccessor:IHttpContextAccessor) =
        ServerSentEventGenerator httpContextAccessor.HttpContext

    member private _.Context = context

    member this.MergeFragment = ServerSentEvent.mergeFragment this.Context EventOptions.defaults
    member this.RemoveFragment = ServerSentEvent.removeFragment this.Context EventOptions.defaults
    member this.MergeStore = ServerSentEvent.mergeStore this.Context EventOptions.defaults
    member this.RemoveFromStore = ServerSentEvent.removeFromStore this.Context EventOptions.defaults
    member this.Redirect = ServerSentEvent.redirect this.Context EventOptions.defaults
    member this.Console = ServerSentEvent.console this.Context EventOptions.defaults

    static member ParseIncoming (httpRequest:HttpRequest) = task {
        let serializedDatastoreOptTask =
            match httpRequest.Method with
            | System.Net.WebRequestMethods.Http.Get ->
                Task.FromResult <|
                    match httpRequest.Query.TryGetValue(Consts.DatastarKey) with
                    | true, json when json.Count > 0 -> ValueSome (json[0])
                    | _ -> ValueNone
            | System.Net.WebRequestMethods.Http.Post ->
                task {
                    try
                        let! readResult = httpRequest.BodyReader.ReadAsync().AsTask()
                        let bytes = readResult.Buffer
                        let str = Encoding.UTF8.GetString(&bytes)
                        return (ValueSome str)
                    with _ -> return ValueNone
                    }
            | _ -> ValueNone |> Task.FromResult
        let! serializedDatastoreOpt = serializedDatastoreOptTask
        return
            try serializedDatastoreOpt |> ValueOption.map (fun serializedDatastore -> JsonSerializer.Deserialize<'T>(serializedDatastore))
            with _ -> ValueNone
            }

    static member ParseIncomingAsync<'T when 'T:null and 'T :> IDatastarStore> (httpRequest:HttpRequest) : Task<'T> = task {
        let! storeVOpt = ServerSentEventGenerator.ParseIncoming(httpRequest)
        return
            match storeVOpt with
            | ValueSome store -> store
            | ValueNone -> null
     }

    with
    interface IServerSentEventGenerator with
        member this.MergeFragment(fragment) = ServerSentEvent.mergeFragment this.Context EventOptions.defaults MergeFragmentOptions.defaults fragment
        member this.MergeFragment(fragment, options) = ServerSentEvent.mergeFragment this.Context EventOptions.defaults options.AsOptions fragment
        member this.MergeFragment(fragment, options, eventOptions) = ServerSentEvent.mergeFragment this.Context eventOptions.AsOptions options.AsOptions fragment
        member this.RemoveFragment(selector) = ServerSentEvent.removeFragment this.Context EventOptions.defaults RemoveFragmentOptions.defaults selector
        member this.RemoveFragment(selector, options) = ServerSentEvent.removeFragment this.Context EventOptions.defaults options.AsOptions selector
        member this.RemoveFragment(selector, options, eventOptions) = ServerSentEvent.removeFragment this.Context eventOptions.AsOptions { SettleDuration = options.SettleDuration; UseViewTransition = options.UseViewTransition } selector
        member this.MergeStore(dataStoreMerge) = ServerSentEvent.mergeStore this.Context EventOptions.defaults false dataStoreMerge
        member this.MergeStore(dataStoreMerge:string, eventOptions:ServerSentEventOpts) = ServerSentEvent.mergeStore this.Context eventOptions.AsOptions false dataStoreMerge
        member this.MergeStore(dataStoreMerge:string, onlyIfMissing:bool) = ServerSentEvent.mergeStore this.Context EventOptions.defaults onlyIfMissing dataStoreMerge
        member this.MergeStore(dataStoreMerge:string, onlyIfMissing:bool, eventOptions:ServerSentEventOpts) = ServerSentEvent.mergeStore this.Context eventOptions.AsOptions onlyIfMissing dataStoreMerge
        member this.RemoveFromStore(dataStorePaths) = ServerSentEvent.removeFromStore this.Context EventOptions.defaults dataStorePaths
        member this.RemoveFromStore(dataStorePaths, eventOptions) = ServerSentEvent.removeFromStore this.Context eventOptions.AsOptions dataStorePaths
        member this.Redirect(uri) = ServerSentEvent.redirect this.Context EventOptions.defaults uri
        member this.Redirect(uri, eventOptions) = ServerSentEvent.redirect this.Context eventOptions.AsOptions uri
        member this.Console(consoleMode, message) = ServerSentEvent.console this.Context EventOptions.defaults consoleMode message
        member this.Console(consoleMode, eventOptions, message) = ServerSentEvent.console this.Context eventOptions.AsOptions consoleMode message

and ServerSentEventHttpContext(httpContext:HttpContext) =
    do
        let setHeader (ctx:HttpContext) (name, content:string) =
            if ctx.Response.Headers.ContainsKey(name) |> not then
                ctx.Response.Headers.Add(name, StringValues(content))
        [
           ("Cache-Control", "no-cache")
           ("Connection", "keep-alive")
           (HeaderNames.ContentType, "text/event-stream")
        ] |> Seq.iter (setHeader httpContext)
        httpContext.Response.StartAsync().GetAwaiter().GetResult()
        httpContext.Response.Body.FlushAsync().GetAwaiter().GetResult()

    member _.Context = httpContext

    with
    interface IServerSentEventContext

    interface ISendEvent with
        member this.SendEvent(event:string) =
            let bytes = Encoding.UTF8.GetBytes(event)
            this.Context.Response.BodyWriter.WriteAsync(bytes).AsTask()

and IServerSentEventGenerator =
    abstract member MergeFragment: fragment:string -> Task
    abstract member MergeFragment: fragment:string * options:MergeFragmentOpts -> Task
    abstract member MergeFragment: fragment:string * options:MergeFragmentOpts * eventOptions:ServerSentEventOpts -> Task
    abstract member RemoveFragment: selector:Selector -> Task
    abstract member RemoveFragment: selector:Selector * options:RemoveFragmentOpts -> Task
    abstract member RemoveFragment: selector:Selector * options:RemoveFragmentOpts * eventOptions:ServerSentEventOpts -> Task
    abstract member MergeStore: dataStore:string -> Task
    abstract member MergeStore: dataStore:string * eventOptions:ServerSentEventOpts -> Task
    abstract member MergeStore: dataStore:string * onlyIfMissing:bool -> Task
    abstract member MergeStore: dataStore:string * onlyIfMissing:bool * eventOptions:ServerSentEventOpts -> Task
    abstract member RemoveFromStore: paths:DataStorePath seq -> Task
    abstract member RemoveFromStore: paths:DataStorePath seq * eventOptions:ServerSentEventOpts -> Task
    abstract member Redirect: url:string -> Task
    abstract member Redirect: url:string * eventOptions:ServerSentEventOpts -> Task
    abstract member Console: consoleMode:ConsoleMode * message:string -> Task
    abstract member Console: consoleMode:ConsoleMode * eventOptions:ServerSentEventOpts * message:string -> Task

and MergeFragmentOpts() =
    member val MergeMode = MergeFragmentOptions.defaults.MergeMode with get, set
    member val Selector = null with get, set
    member val SettleDuration = MergeFragmentOptions.defaults.SettleDuration with get, set
    member val UseViewTransition = MergeFragmentOptions.defaults.UseViewTransition with get, set
    member this.AsOptions =
        { MergeMode = this.MergeMode
          Selector = Utility.ValueOption.fromNullable this.Selector
          SettleDuration = this.SettleDuration
          UseViewTransition = this.UseViewTransition }

and RemoveFragmentOpts() =
    member val SettleDuration = MergeFragmentOptions.defaults.SettleDuration with get, set
    member val UseViewTransition = MergeFragmentOptions.defaults.UseViewTransition with get, set
    member this.AsOptions =
        { SettleDuration = this.SettleDuration
          UseViewTransition = this.UseViewTransition }

and ServerSentEventOpts() =
    member val EventId = null with get, set
    member val Retry = Consts.DefaultSSERetryDuration with get, set
    member this.AsOptions =
        { EventId = Utility.ValueOption.fromNullable this.EventId
          Retry = this.Retry }
